### ISSUE TITLE
Make `cholesky(Positive, A)` differentiable 

### ISSUE THREAD

**Issue by bvdmitri** [timholy/PositiveFactorizations.jl#26]
Hi!

Thank you for your package!

Would it be possible to make `cholesky(Positive, x)` differentiable with `ForwardDiff.jl`? 

MWE:

```Julia
using PositiveFactorizations
using ForwardDiff
using LinearAlgebra

g = (x) -> ForwardDiff.gradient((x) -> det(inv(cholesky(x))), x)
f = (x) -> ForwardDiff.gradient((x) -> det(inv(cholesky(Positive, x))), x)

A = Matrix(Diagonal(ones(3)))

g(A) # ok
f(A) # Fails with `PositiveFactorizations.floatype` not defined
```

Adding an extra method to `PositiveFactorizations.floattype` doesn't really help, since `ldlt!` function accepts only `AbstractFloat` type, while `ForwardDiff.Dual` is `Real`. 

```Julia
PositiveFactorizations.floattype(::Type{ T }) where { T <: ForwardDiff.Dual } = T

f(A)
# 
MethodError: no method matching ldlt!(::Type{Positive{ForwardDiff.Dual{ForwardDiff.Tag{var"#65#67",Float64},Float64,9}}}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{var"#65#67",Float64},Float64,9},2}, ::Type{Val{false}}; tol=Dual{ForwardDiff.Tag{var"#65#67",Float64}}(6.661338147750939e-15,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,6.661338147750939e-15), blocksize=10)
```

Furthermore we can relax type requirements to be `<: Real`, also some checks in the original code actually check eltype of a matrix to be `<: Real`. But then it fails on [this line of code](https://github.com/timholy/PositiveFactorizations.jl/blob/7b5c4458bb9718c25dc381442344699df2891a2c/src/cholesky.jl#L83) due to the fact that `d` array has a predefined eltype `Int8`.

**Comment by timholy** [comment]
For performance this calls BLAS, so it won't work. I'd be happy to accept a PR providing a non-BLAS alternative. But breaking the BLAS interface wouldn't be accepted, because people who don't need differentiability get a lot of benefit from the speed.

**Comment by bvdmitri** [comment]
@timholy Thank you for fast reply!

I see. I quickly checked the source code of the package and it seems to me that the only one place where BLAS routines are used is in `update_columns!` function. And it already has non-BLAS fallback [here](https://github.com/timholy/PositiveFactorizations.jl/blob/master/src/cholesky.jl#L207) and [here](https://github.com/timholy/PositiveFactorizations.jl/blob/master/src/cholesky.jl#L231). I believe builtin `cholesky` from `LinearAlgebra` works in the same way (and it is differentiable).

I think the main problem with differentiability of `cholesky(Positive, A)` is the assumption that `sign` function always returns integer [here](https://github.com/timholy/PositiveFactorizations.jl/blob/master/src/cholesky.jl#L25). For `Dual` it is not true. How do you think would it be possible to drop this assumption in the code?

**Comment by bvdmitri** [comment]
@timholy Can we introduce a new function `signtype(::Type)`? 

Something like:

```
function ldlt!(...)
    ...
    d = Array{signtype(T)}(undef, K)
    ...
end

signtype(::Type{T}) where {T<:AbstractFloat} = Int8
signtype(::Type{T}) where {T<:Integer} = Int8
```

I've tested it in my fork and with additional
```
import ForwardDiff
import PositiveFactorizations

PositiveFactorizations.floattype(::Type{ T }) where { T <: ForwardDiff.Dual } = T
PositiveFactorizations.signtype(::Type{ T }) where { T <: ForwardDiff.Dual } = T
```
`cholesky(Positive, A)` becomes differentiable.

I can make a PR if you agree with this change.

**Comment by timholy** [comment]
Definitely!

**Comment by timholy** [comment]
Closed by #27
